### PR TITLE
Unique network flows

### PR DESCRIPTION
**Pull request by anamud**
This PR implements unique network flows for publishers and subscribers in communicating nodes. 

Associated pull-requests:
- [design](https://github.com/ros2/design/pull/304)
- [examples](https://github.com/ros2/examples/pull/296)
- [rcl](https://github.com/ros2/rcl/pull/880)
- [rmw](https://github.com/ros2/rmw/pull/294)
- [rmw_implementation](https://github.com/ros2/rmw_implementation/pull/170)
- [rmw_fastrtps](https://github.com/ros2/rmw_fastrtps/pull/502)
- [rmw_cyclonedds](https://github.com/ros2/rmw_cyclonedds/pull/282)
- [rmw_connext](https://github.com/ros2/rmw_connext/pull/484) (NB: This repository is soon to be deprecated)
- [rmw_connextdds](https://github.com/ros2/rmw_connextdds/pull/13)
- [Fast DDS](https://github.com/eProsima/Fast-DDS/pull/1682)

Initial contributions stem from Ericsson and eProsima.

Signed-off-by: Ananya Muddukrishna ananya.x.muddukrishna@ericsson.com

### CODE REVIEW

**Comment by anamud** [comment]
@fujitatomoya @ivanpauno Kindly review the PR and suggest improvements if required.

File: rclcpp/include/rclcpp/network_flow_endpoint.hpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  {
+  }
+
+  /// Get transport protocol
+  RCLCPP_PUBLIC
+  const std::string & transport_protocol() const;
+
+  /// Get internet protocol
+  RCLCPP_PUBLIC
+  const std::string & internet_protocol() const;
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/network_flow_endpoint.hpp:58]: I wonder, should these be stored as fixed sized c-strings (or maybe a string_view to a c-string) and returned as `const char *` (or again as string_view)?

They never change and their maximum size is already known, so it seems like `std::string` is overkill?

File: rclcpp/include/rclcpp/network_flow_endpoint.hpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  RCLCPP_PUBLIC
+  uint32_t flow_label() const;
+
+  /// Get DSCP
+  RCLCPP_PUBLIC
+  uint8_t dscp() const;
+
+  /// Get internet address
+  RCLCPP_PUBLIC
+  const std::string & internet_address() const;
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/network_flow_endpoint.hpp:74]: Same thought here.

File: rclcpp/src/rclcpp/publisher_base.cpp [hunk 3]
```diff
@@ -268,3 +269,40 @@ PublisherBase::default_incompatible_qos_callback(
...
+    }
+    rclcpp::exceptions::throw_from_rcl_error(ret, error_msg);
+  }
+
+  std::vector<rclcpp::NetworkFlowEndpoint> network_flow_endpoint_vector;
+  for (size_t i = 0; i < network_flow_endpoint_array.size; ++i) {
+    network_flow_endpoint_vector.push_back(
+      rclcpp::NetworkFlowEndpoint(
+        network_flow_endpoint_array.
+        network_flow_endpoint[i]));
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/publisher_base.cpp:299]: nitpick: This should be ok right? The wrap seems unnecessary:

```suggestion
        network_flow_endpoint_array.network_flow_endpoint[i]));
```

**Review by wjwwood — APPROVED** [review]
lgtm, with small comments and suggestions

File: rclcpp/src/rclcpp/publisher_base.cpp (hunk 3 above)
Comment (anamud) [reply rclcpp/src/rclcpp/publisher_base.cpp]: Fixed. See [here](https://github.com/ros2/rclcpp/pull/1496/commits/31d418afc1b1956c52e4d7ed3c8b36be266abca9).

File: rclcpp/include/rclcpp/network_flow_endpoint.hpp (hunk 1 above)
Comment (anamud) [reply rclcpp/include/rclcpp/network_flow_endpoint.hpp]: Thought about this and tried a couple of approaches -- guarding the member string as const and using a char buffer. Both made me uneasy. The constant member string upset move semantics. The char buffer became unwieldy due to unclear max size and increased coupling with rmw. 

I prefer to keep it as std::string and relay on optimizations (e.g. SBO) by the compiler to retain speed. The interface is also not on the typical critical path and is most likely going to be used once during the lifetime of a publisher/subscription. 

When C++17 becomes the norm for ROS2 packages, the std::string container can be switched to std::string_view.

File: rclcpp/include/rclcpp/network_flow_endpoint.hpp (hunk 2 above)
Comment (anamud) [reply rclcpp/include/rclcpp/network_flow_endpoint.hpp:67]: See previous reply [here](https://github.com/ros2/rclcpp/pull/1496#discussion_r606702526).

File: rclcpp/include/rclcpp/network_flow_endpoint.hpp (hunk 1 above)
Comment (wjwwood) [reply rclcpp/include/rclcpp/network_flow_endpoint.hpp:73]: Ok, technically that's already the case (we're using `std::variant` now), but we can wait.

**Review by wjwwood — APPROVED** [review]

**Pull request merged**